### PR TITLE
Add utmpx

### PR DESCRIPTION
**Pull request by knight42**
#338

### CODE REVIEW

File: src/unix/notbsd/linux/mod.rs [hunk 1]
```diff
@@ -481,6 +506,23 @@ pub const RB_POWER_OFF: ::c_int = 0x4321fedcu32 as i32;
...
+pub const BOOT_TIME: ::c_short = 2;
+pub const NEW_TIME: ::c_short = 3;
+pub const OLD_TIME: ::c_short = 4;
+pub const INIT_PROCESS: ::c_short = 5;
+pub const LOGIN_PROCESS: ::c_short = 6;
+pub const USER_PROCESS: ::c_short = 7;
+pub const DEAD_PROCESS: ::c_short = 8;
+pub const ACCOUNTING: ::c_short = 9;
+pub const _PATH_UTMP: &'static str = "/var/run/utmp";
+pub const _PATH_WTMP: &'static str = "/var/log/wtmp";
```
Comment (alexcrichton) [inline src/unix/notbsd/linux/mod.rs]: Currently we don't include constant strings like this in libc, perhaps these could be omitted for now?

File: src/unix/notbsd/linux/mod.rs [hunk 2]
```diff
@@ -192,6 +192,31 @@ s! {
...
+        pub ut_type: ::c_short,
+        pub ut_pid: ::pid_t,
+        pub ut_line: [::c_char; UT_LINESIZE],
+        pub ut_id: [::c_char; UT_IDSIZE],
+
+        pub ut_user: [::c_char; UT_NAMESIZE],
+        pub ut_host: [::c_char; UT_HOSTSIZE],
+        pub ut_exit: exit_status,
+        pub ut_session: ::int32_t,
+        pub ut_tv: _timeval,
```
Comment (alexcrichton) [inline src/unix/notbsd/linux/mod.rs]: Out of curiosity, is this equivalent to `timeval`?

Comment (knight42) [inline src/unix/notbsd/linux/mod.rs]: Not always.
```
pub struct timeval {
    pub tv_sec: time_t (AKA i64 on 64-bit platform),
    pub tv_usec: suseconds_t (AKA i64 on 64-bit platform),
}
```

However, according the definition of `struct utmp` on linux:
```C
struct utmp {
    ...
    ...

    /* The ut_session and ut_tv fields must be the same size when
       compiled 32- and 64-bit.  This allows data files and shared
       memory to be shared between 32- and 64-bit applications. */
    #if __WORDSIZE == 64 && defined __WORDSIZE_COMPAT32
    int32_t ut_session;           /* Session ID (getsid(2)),
                                     used for windowing */
    struct {
        int32_t tv_sec;           /* Seconds */
        int32_t tv_usec;          /* Microseconds */
    } ut_tv;                      /* Time entry was made */
    #else
    long   ut_session;           /* Session ID */
    struct timeval ut_tv;        /* Time entry was made */
    #endif

    ...
};
```
and it seems that `__WORDSIZE_COMPAT32` is usually defined on 64-bit platform, so the actual definition of `struct utmp` goes like this:
```C
struct utmp {
    ...
    ...

    int32_t ut_session;           /* Session ID (getsid(2)),
                                     used for windowing */
    struct {
        int32_t tv_sec;           /* Seconds */
        int32_t tv_usec;          /* Microseconds */
    } ut_tv;                      /* Time entry was made */

    ...
};
```

It's worth noting that getutent() reads `struct utmp`s from binary file /var/run/utmp, hence the actual size of `struct utmp` matters.

File: src/unix/notbsd/linux/mod.rs (hunk 1 above)
Comment (knight42) [inline src/unix/notbsd/linux/mod.rs]: Removed.

File: src/unix/notbsd/linux/mod.rs (hunk 2 above)
Comment (alexcrichton) [inline src/unix/notbsd/linux/mod.rs]: Ah oh well, thanks for the clarification though!

**Comment by alexcrichton** [comment]
Looks like the CI is failing for non-spurious reasons?

**Comment by knight42** [comment]
Yep, according to the stderr, cc could not find the definition of the structs I added. Trying to find out why.

**Comment by knight42** [comment]
@alexcrichton Got it. `libc-test` doesn't include `utmp.h`.

**Comment by knight42** [comment]
@alexcrichton Some irrelated errors happened this time... 

Anyway, I decided to compromise, only add `struct utmp(x)` on certain platforms.

**Comment by alexcrichton** [comment]
@bors: r+ b47e554216fadb9613f4ff8ec271ecd9a5bcbc7f

Thanks!

**Comment by alexcrichton** [comment]
grr bors, but tests passed!

**Pull request merged**